CORDA-3201 - Enforce separate key for notary identity

oknowles opened the pull request:
# Overview
This PR modifies the initial registration logic to enfore all OS notary implementations to use a separate identity/key to sign notarisation requests. When a new notary node is registered a separate identity is generated as well as the standard 'node CA' and 'node legal identity'.

The result is that a notary node info contains two identities, with the below "O=Notary Service..." becoming the identity that is associated with transactions:
```
>>> run networkMapSnapshot
- addresses:
  - "localhost:10002"
  legalIdentitiesAndCerts:
  - "O=Notary Node, L=Zurich, C=CH"
  - "O=Notary Service, L=Zurich, C=CH"
  platformVersion: 7
  serial: 1591177242227
...
```

# Changes
The logic for loading and using the correct notary identity when constructing the notary services was already present due to the experimental RAFT and BFT notaries. Therefore the bulk of the changes just ensure that the separate notary identity is generated during initial registration. The main changes are:
* Enforce `notary.serviceLegalName` config parameter present within the initial registration logic.
* Add logic to generate the notary service identity prior to performing standard initial registration for a notary node.
* Update samples to include a separate notary service identity for each notary node.

# Backwards Compatibility
The `notary.serviceLegalName` config parameter presence is only enforced within the registration logic. The logic to load the notary identity is unchanged, and will load the node's legal identity if no service identity has been defined. Therefore older notaries can be upgraded to a new Corda version and continue to sign requests using the same key, but new notaries cannot be created using the old identity structure.

# Samples & Network Bootstrapper
Although the samples have been updated to include a separate notary service identity, the `deployNodes`/`runNodes` gradle tasks will still work without the `notary.serviceLegalName` config addition. This is because the nodes are not initialised using the same initial registration logic that a production like network would use. The samples have still been updated however to familiarise the users with the config for when they deploy an actual network.

The network bootstrapper is similar in this respect. Due to the previously existing RAFT and BFT notaries, adding in the `notary.serviceLegalName` for a simple notary will result in the separate notary identity being generated, but networks will still work without this option.

# Testing
Alongside the addition of unit tests, a local network has been created using both the `deployNodes` gradlew task and via the standard, production like network setup.

Commit:
Enforce separate notary identity during initial registration.

Commit:
Add to network registration tests to validate notary registration logic

Commit:
Remove temporary hardcoding of truststore password during notary registration

Commit:
minor refactoring

Commit:
Small clarification in comment relating to notary config validation

Commit:
Update deployNodes gradle tasks in samples to include notary service legal name

Commit:
Renaming and commenting around loading of notary identity to provide more clarity

Commit:
Update detekt baseline

thschroeter commented on common/logging/src/main/kotlin/net/corda/common/logging/Constants.kt:
```diff
@@ -9,4 +9,4 @@ package net.corda.common.logging
  */
 
-internal const val CURRENT_MAJOR_RELEASE = "4.6-SNAPSHOT"
+internal const val CURRENT_MAJOR_RELEASE = "4.6-SNAPSHOT"
```
a spurious newline change

thschroeter commented on node/src/main/kotlin/net/corda/node/utilities/registration/NetworkRegistrationHelper.kt:
```diff
@@ -287,10 +321,29 @@ class NodeRegistrationConfiguration(
+                // exists. Just having the validation here prevents any new notaries from being created with the legacy identity scheme
+                // but still allows drop in JAR replacements for old notaries.
+                requireNotNull(it.serviceLegalName) {
+                    "The notary service legal name must be provided"
```
Perhaps we could mention  that this is done with the `notary.serviceLegalName` config entry.

thschroeter reviewed (commented):
Thanks @oknowles 

Should we add a check that the notary service legal name is present whenever there is a `distributed-notary-service-id` in the keystore?

I'm worried about the case when the operator removes the service legal name from the config and the notary effectively leaves the cluster because it'll generate a new node info without the service legal ID and cert of the service. It will not receive requests any more but if it does it would start signing with the regular key.

Commit:
Add check during notary startup to validate if notary service key exists then notary service legal name is present

Commit:
Address PR comments

oknowles commented:
> Should we add a check that the notary service legal name is present whenever there is a `distributed-notary-service-id` in the keystore?
> 
> I'm worried about the case when the operator removes the service legal name from the config and the notary effectively leaves the cluster because it'll generate a new node info without the service legal ID and cert of the service. It will not receive requests any more but if it does it would start signing with the regular key.

This is a very good point. I have added in a check during node startup (when the notary identity key is loaded) to validate this behaviour. 

I performed a quick test of this scenario to confirm that the appropriate error would be returned. That is, I registered the notary with a separate service identity and removed the `notary.serviceLegalName` configuration parameter before starting it. The below screenshot shows the desired outcome (and obviously adding the `notary.serviceLegalName` configuration parameter back in works as usual):

<img width="829" alt="Screenshot 2020-06-17 at 11 54 32" src="https://user-images.githubusercontent.com/12070000/84889870-7cdfa700-b091-11ea-8bba-402fc6e36027.png">

Commit:
Revert changes to example build.gradle file as it was removed in release branch

thschroeter reviewed (approved):
LGTM

rekalov reviewed (approved):
LGTM
